### ISSUE TITLE
Standard traits exception handler `_log_exception` should not blindly create handler

### ISSUE THREAD

**Issue by burnpanck** [enthought/traits#265]
I'm referring to the method `_log_exception` of the class `traits.trait_notifiers.NotificationExceptionHandler`, which will add `logging.StreamHandler` to the list of the handlers managed by the `logging` library. This makes it difficult for users of `traits` to keep stdout/stderr free from exception messages. Instead, it should either just not add a logging handler, or at least check if any other handler is registered.

**Comment by mdickinson** [comment]
Agreed.  I think we should simply remove that code.

**Comment by mdickinson** [comment]
> I think we should simply remove that code.

Though perhaps it's not quite that simple.  Unnoticed exceptions in traits handlers are a common cause of hard-to-find bugs.  If there's something we can do to prevent them going unnoticed in the case where a Traits user hasn't set up any logging (or alert that user that (s)he should look at his logs in the case where logging *has* been set up), that would be helpful.

**Comment by rkern** [comment]
It looks like the we would get the same intended result by logging the exception (so it gets picked up by other handlers added elsewhere) and then simply printing the traceback to `sys.stderr`.

**Comment by burnpanck** [comment]
@rkern: That would defeat the purpose of my proposal: The preventing unwanted output in the standard streams unless chosen by the user!
@mdickinson: The function `logging.basicConfig` is the standard library approach to preventing error messages unnoticed [(see library documentation)](https://docs.python.org/2/library/logging.html#logging.basicConfig): It installs a handler only if none is installed yet. Maybe that should be used?

**Comment by rkern** [comment]
There is always a `NullHandler` installed for the `traits` logger, as is best practice for libraries. Nothing other than `main()` functions should ever call `logging.basicConfig()`.

Perhaps we should call `sys.excepthook()` here, as is standard for other things that face this problem, like GUI event loops. General-purpose exception handlers like this, that for whatever reason don't or can't reraise the exception, need to be noisy by default, or else they hide bugs. You should be able to opt *out* of that noisiness, and `sys.excepthook()` is a standard way of doing so. Requiring that users configure things to opt *in* to getting these messages hides bugs. We have had *lots* of experience with Traits hiding exceptions. The precise details of how we make sure these exceptions get reported by default, and you are right that the current implementation is flawed, but we really do need to report them by default one way or the other.

**Comment by burnpanck** [comment]
I guess you are right with the opt-out. I would prefer if I could opt-out generally using a standard mechanism, rather than a specialised mechanism for every library. I think that mechanism should be the logging library. I don't know about `sys.excepthook` I thought that is for exceptions that cannot be raised. In the case of the traits standard error handler, the exception will get re-raised. The logging it is doing is not happening because it cannot raise for any reason, it is just in addition logging. Now the problem is how to make the `logging` library behave as opt-out. (It should offer this function, as this problem is common to any logging of serious issues). I thought that `basicConfig` is exactly this: It installs a stream handler unless the user opts-out by installing something himself. Are you sure that it should not be called except from `main`? (In the docs, there is a note about a race-condition in earlier versions of python). For example, all of `logging.info`, `logging.warn` and `logging.exception` do that internally according to the docs... Note that the opt-out is based only on the state of the `root` logger's handlers not anything else. I would have thought that this is more or less the opt-out that one is expecting...

**Comment by rkern** [comment]
The exception *isn't* being reraised. That's the only reason we are logging or printing anything about it.

**Comment by rkern** [comment]
As for using a standard method to keep your `sys.stderr` clear, then `sys.excepthook()` is it, as you will have to set that to deal with exceptions that happen inside your GUI event loop.

**Comment by mdickinson** [comment]
No, we definitely don't want to use `logging.basicConfig` within the Traits library code.  One reason is that it could prevent applications from being able to usefully use `logging.basicConfig` themselves: if Traits has already done a `logging.basicConfig()` (perhaps at import time) before the application sets up logging, the application's own `logging.basicConfig(**carefully_chosen_arguments)` [has no effect](https://docs.python.org/2/library/logging.html?highlight=basicconfig#logging.basicConfig), and then the app developer wonders why none of his log messages are showing up in the right place.  (For the same reason, we shouldn't be using `logging.info` and friends within Traits, either; any place we're doing so is a bug, and should be fixed.)

**Comment by burnpanck** [comment]
You are right, it is not re-raised. How did I come to the other conclusion? I remember seeing doubled trace-backs, one on stderr, one on stdout. I somehow concluded that one comes from my ipython prompt displaying the exception it gets and one from traits... However, that must have had another cause (and wasn't the reason why I filed this issue: I just noticed during writing of #266). Of course, when raising is not an option, then I agree, `sys.excepthook` is the right solution. Sorry for my stubbornness!
